### PR TITLE
[Serve] Add Tutorial for Batch Inference

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
Add an example and tutorial for serving batch queries.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: doc/source/rayserve/tutorials/batch-tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,119 @@
+.. _serve-batch-tutorial:
+
+Batching Tutorial
+=================
+
+In this guide, we will deploy a simple vectorized adder that takes
+a batch of queries and add them at once. In particular, we show:
```
Comment (deanwampler) [inline doc/source/rayserve/tutorials/batch-tutorial.rst:7]: "adds them at once".

File: doc/source/rayserve/tutorials/batch-tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  *utilization* of the hardware.
+
+
+Let's import RayServe and some other helpers.
+
+.. literalinclude:: ../../../../python/ray/serve/examples/doc/tutorial_batch.py
+    :start-after: __doc_import_begin__
+    :end-before: __doc_import_end__
+
+You can use ``@serve.accept_batch`` decorator to annotate a function or a class.
```
Comment (deanwampler) [inline doc/source/rayserve/tutorials/batch-tutorial.rst:29]: "You can use the "... "compared to single request backends".

File: doc/source/rayserve/tutorials/batch-tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        pass
+
+For batched backend, the inputs types are converted to list of their original
+types:
+
+.. code-block:: python
+
+    @serve.accept_batch
+    def blacked_request(
+        flask_request: List[Flask.Request],
```
Comment (deanwampler) [inline doc/source/rayserve/tutorials/batch-tutorial.rst:52]: Meta comment. What if you had just one `*_request` method and it always takes a list, even for single requests, where the list will have one item. A bit wasteful for single requests, but it simplifies everything else.

File: doc/source/rayserve/tutorials/batch-tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    ):
+        pass
+
+For batched backend, the inputs types are converted to list of their original
+types:
+
+.. code-block:: python
+
+    @serve.accept_batch
+    def blacked_request(
```
Comment (deanwampler) [inline doc/source/rayserve/tutorials/batch-tutorial.rst:51]: Did you mean `batched_request`?

File: doc/source/rayserve/tutorials/batch-tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+Let's define the backend function. We will take in a list of requests, extract
+the input value, convert them into an array, and use NumPy to add 1 to each element.
+
+.. literalinclude:: ../../../../python/ray/serve/examples/doc/tutorial_batch.py
+    :start-after: __doc_define_servable_v0_begin__
+    :end-before: __doc_define_servable_v0_end__
+
+Let's deploy it. Note that in the ``config`` section of ``create_backend``, we
+are specifying the maximum batch size via ``config={"max_batch_size": 4}``. This
+configuration option limit the maximum possible batch size send to the backend.
```
Comment (deanwampler) [inline doc/source/rayserve/tutorials/batch-tutorial.rst:67]: "limits"... "sent to the backend"

File: doc/source/rayserve/tutorials/batch-tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    ``max(number_of_pending_queries, max_batch_size)`` queries to form a batch.
+
+.. literalinclude:: ../../../../python/ray/serve/examples/doc/tutorial_batch.py
+    :start-after: __doc_deploy_begin__
+    :end-before: __doc_deploy_end__
+
+Let's define a :ref:`Ray remote task<ray-remote-functions>` to send queries in
+parallel. As you can see, the first batch has a batch size of 1, and the subsequent
+queries have a batch size of 4. Even though each query is issued independently,
+the RayServe was able to evaluate them in batches.
```
Comment (deanwampler) [inline doc/source/rayserve/tutorials/batch-tutorial.rst:81]: Drop "the" before Ray Serve.

File: doc/source/rayserve/tutorials/batch-tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+.. literalinclude:: ../../../../python/ray/serve/examples/doc/tutorial_batch.py
+    :start-after: __doc_query_begin__
+    :end-before: __doc_query_end__
+
+What if you want to evaluate a whole batch in Python? RayServe allows you to send
+queries via Python API. You can use the boolean value ``serve.context.web`` to
```
Comment (deanwampler) [inline doc/source/rayserve/tutorials/batch-tutorial.rst:88]: "via a Python API" (or maybe "the Python API")

File: doc/source/rayserve/tutorials/batch-tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+distinguish the origin of the queries. A batch of queries can either come from
+the web server or the Python API. RayServe will guarantee there won't be queries
+with mixed origins.
+
+When the batch of requests come from the web API, RayServe will fill the first
```
Comment (deanwampler) [inline doc/source/rayserve/tutorials/batch-tutorial.rst:93]: "comes" (think of it like this: "When the batch comes")

File: doc/source/rayserve/tutorials/batch-tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+argument ``flask_requests`` with a list of `Flask.Request` object and set
```
Comment (deanwampler) [inline doc/source/rayserve/tutorials/batch-tutorial.rst:94]: "objects"

File: doc/source/rayserve/tutorials/batch-tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+``serve.context.web = True``. When the batch of requests come from the Python API,
```
Comment (deanwampler) [inline doc/source/rayserve/tutorials/batch-tutorial.rst:95]: "comes"

**Review by deanwampler — APPROVED** [review]
Looks good. Most comments are spelling or grammar corrections.

File: doc/source/index.rst [hunk 11]
```diff
@@ -286,6 +286,7 @@ Getting Involved
    rayserve/tutorials/tensorflow-tutorial.rst
    rayserve/tutorials/pytorch-tutorial.rst
    rayserve/tutorials/sklearn-tutorial.rst
+   rayserve/tutorials/batch-tutorial.rst
```
Comment (anabranch) [inline doc/source/index.rst:289]: I guess I don't totally understand why this needs to be broken out into its own page.

Alternatively, couldn't we break out a section called "Backend Configurations" or something like that and have all of these here? Also, we have batching [here](https://docs.ray.io/en/master/rayserve/overview.html#batching), is this something different? Is it just a more in depth walkthrough of batching?

Comment (simon-mo) [reply doc/source/index.rst:289]: This is a learning-oriented tutorial. The batching in over page is goal-oriented how-to guides.

File: doc/source/serve/tutorials/batch.rst [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    :end-before: __doc_query_end__
+
+What if you want to evaluate a whole batch in Python? Ray Serve allows you to send
+queries via the Python API. You can use the boolean value ``serve.context.web`` to
+distinguish the origin of the queries. A batch of queries can either come from
+the web server or the Python API. Ray Serve will guarantee there won't be queries
+with mixed origins.
+
+When the batch of requests comes from the web API, Ray Serve will fill the first
+argument ``flask_requests`` with a list of `Flask.Request` objects and set
```
Comment (anabranch) [inline doc/source/serve/tutorials/batch.rst:94]: `Flask.Request` => ``Flask.Request``

File: doc/source/serve/tutorials/batch.rst [hunk 13]
```diff
@@ -0,0 +1,119 @@
+.. _serve-batch-tutorial:
```
Comment (anabranch) [inline doc/source/serve/tutorials/batch.rst:1]: Will you please link this from the batching bit in the [advanced](https://docs.ray.io/en/master/serve/advanced.html#batching-to-improve-performance) section?

**Review by bllchmbrs — APPROVED** [review]
LGTM Except that I would like you to add a link to this - great job!

**Pull request merged**